File: Data Structures & Algorithms/three-integer-sum/submission-2.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        // 结果列表
        List<List<Integer>> result = new ArrayList<>();
        
        // 排序数组
        Arrays.sort(nums);
        
        // 遍历数组，固定第一个数 nums[i]
        for (int i = 0; i < nums.length - 2; i++) {
            // 如果当前数字与前一个数字相同，跳过以避免重复结果
            if (i > 0 && nums[i] == nums[i - 1]) {
                continue;
            }
            
            // 使用双指针查找两个数，使得三数之和为零
            int left = i + 1; // 左指针初始化为 i 之后的第一个数
            int right = nums.length - 1; // 右指针初始化为数组末尾
            
            while (left < right) {
                int sum = nums[i] + nums[left] + nums[right];
                
                // 如果三数之和为零，将结果加入列表
                if (sum == 0) {
                    result.add(Arrays.asList(nums[i], nums[left], nums[right]));
                    
                    // 移动左指针，跳过重复元素
                    while (left < right && nums[left] == nums[left + 1]) {
                        left++;
                    }
                    
                    // 移动右指针，跳过重复元素
                    while (left < right && nums[right] == nums[right - 1]) {
                        right--;
                    }
                    
                    // 移动指针，寻找新的组合
                    left++;
                    right--;
                } else if (sum < 0) {
                    // 如果三数之和小于零，移动左指针使和增大
                    left++;
                } else {
                    // 如果三数之和大于零，移动右指针使和减小
                    right--;
                }
            }
        }
        
        // 返回结果列表
        return result;
    }
}//on^2,on
//15

File: Data Structures & Algorithms/palindromic-substrings/submission-2.py
class Solution:
    def countSubstrings(self, s: str) -> int:
        N = len(s)
        if N == 0:
            return 0

        #dp = [[False] * N for _ in range(N)]
        dp = []                 # 先准备一个空列表（大盒子）
        for i in range(N):      # 循环 N 次，创建 N 行
            row = []            # 当前这一行（小盒子）
            
            for j in range(N):  # 每一行放 N 个 False
                row.append(False)

            dp.append(row)      # 把这一行放进 dp

        result = 0

        for L in range(1, N + 1):
            for i in range(N - L + 1):
                j = i + L - 1

                if L == 1:
                    dp[i][j] = True
                elif L == 2:
                    dp[i][j] = (s[i] == s[j])
                else:
                    dp[i][j] = (s[i] == s[j]) and dp[i + 1][j - 1]

                if dp[i][j]:
                    result += 1

        return result
